File: src/202_happy_number/happy_number.py
# leetcode/src/202_happy_number/happy_number.py
# Skeleton for LeetCode 202. Happy Number
# Follow-up implementations should replace the `pass` in Solution.isHappy
# Test scaffolding prints: actual | expected | PASS/FAIL

from typing import List


class Solution:
    def isHappy(self, n: int) -> bool:
        """
        Determine if a number n is a happy number.

        Uses Floyd's cycle detection (tortoise and hare). Ensure we check
        for the accepting state (1) before deciding that a meeting point
        indicates a non-happy cycle (avoids false negative when both
        pointers reach 1 in the same iteration).
        """
        slow, fast = n, n

        while True:
            slow = self.sumOfSquares(slow)
            fast = self.sumOfSquares(self.sumOfSquares(fast))

            if slow == 1 or fast == 1:
                return True

            if slow == fast:
                return False

    def sumOfSquares(self, n: int) -> int:
        res = 0
        while n > 0:
            num = n % 10
            res += num**2
            n //= 10
        return res


def run_tests() -> None:
    solution = Solution()

    test_cases = [
        (19, True),
        (2, False),
        (1, True),
        (7, True),
        (116, False),
        (100, True),
    ]

    for n, expected in test_cases:
        actual = solution.isHappy(n)
        status = "PASS" if actual == expected else "FAIL"
        print(f"{actual} | {expected} | {status}")


if __name__ == "__main__":
    run_tests()

File: src/202_happy_number/happy_number.cpp
#include <iostream>
#include <string>
#include <vector>
using namespace std;

// C++ implementation for LeetCode 202 - Happy Number
// Uses Floyd's cycle detection (tortoise and hare) and includes
// test scaffolding that prints results in the format:
// actual | expected | PASS (if actual == expected) or FAIL

class Solution {
public:
    // Determine whether n is a happy number.
    bool isHappy(int n) {
        if (n <= 0) return false;

        auto sumOfSquares = [](int x) {
            int res = 0;
            while (x > 0) {
                int d = x % 10;
                res += d * d;
                x /= 10;
            }
            return res;
        };

        int slow = sumOfSquares(n);
        int fast = sumOfSquares(sumOfSquares(n));

        while (slow != fast) {
            slow = sumOfSquares(slow);
            fast = sumOfSquares(sumOfSquares(fast));
        }

        return slow == 1;
    }
    private:
    int sumOfSquares(int n) {
        int res = 0;
        while (n > 0) {
            int digit = n % 10;
            res += digit * digit;
            n /= 10;
        }

        return res;
    }
};

// Helper to format boolean output as "true"/"false"
static string bool_to_str(bool b) {
    return b ? "true" : "false";
}

// Test scaffolding with 5+ test cases.
// Prints results in the required format:
// actual | expected | PASS (if actual == expected) or FAIL
int main() {
    Solution solution;

    vector<pair<int, bool>> tests = {
        {19, true},    // example: happy
        {2, false},    // example: not happy
        {1, true},     // trivial happy
        {7, true},     // known happy number
        {116, false},  // known non-happy
        {100, true},   // additional case
        {4, false}     // known unhappy cycle starter
    };

    for (const auto &tc : tests) {
        int input = tc.first;
        bool expected = tc.second;

        bool actual = solution.isHappy(input);

        cout << bool_to_str(actual)
             << " | " << bool_to_str(expected)
             << " | " << (actual == expected ? "PASS" : "FAIL")
             << "\n";
    }

    return 0;
}
